improve diagnostic for unnamed closure parameters

AppAppWorks opened the pull request:
fix #2340

AppAppWorks commented:
@bnbarham would you mind reviewing this PR too? :)

rintaro commented on Sources/SwiftParser/Parameters.swift:
```diff
@@ -160,11 +160,20 @@ extension Parser {
+    var colon = self.consume(if: .colon)
+    // try to parse the type regardless of the presence of the preceding colon
+    // because the parameter may be unnamed
+    // e.g. { [X] in } or { (:[X]) in }
```
Are these relevant here? From the code of `parseParameterClause`, I feel like `parseClosureParameter` is not even called.
If it does, could you add test cases for these patterns?

AppAppWorks replied:
I may just trim the braces from the examples in the comments.

rintaro commented on Sources/SwiftParser/Parameters.swift:
```diff
@@ -160,11 +160,20 @@ extension Parser {
+    // try to parse the type regardless of the presence of the preceding colon
+    // because the parameter may be unnamed
+    // e.g. { [X] in } or { (:[X]) in }
+    var type: RawTypeSyntax? = self.parseType(misplacedSpecifiers: misplacedSpecifiers)
```
I feel this a bit scary. Could you `Lookahead.canParseType()` and check if the next token is `,` or `)`, instead of checking `.missingType`.

AppAppWorks replied:
I wasn't aware of the existence of `Lookahead` until earlier today, sorry for that ;(

rintaro commented on Tests/SwiftParserTest/translated/RecoveryTests.swift:
```diff
@@ -3355,4 +3355,19 @@ final class RecoveryTests: ParserTestCase {
+    // <rdar://1181099123>
+    assertParse(
+      """
+      test { (1️⃣[X]) in }
```
Do we have a test case for missing label but with colon? e.g. `test { (: FooBar) in }`

bnbarham replied:
With an attribute would also be interesting, ie. `test { (foo @bar baz) in }`.

Also note that all the `translated/*` tests were translated from the compiler parser tests. It'd probably be better to have this one appropriately named in `ExpressionTests` (though in an ideal world we'd split that file up too).

AppAppWorks replied:
> Do we have a test case for missing label but with colon? e.g. `test { (: FooBar) in }`

For some reason I deleted that test case from the final commit, I'll add it back.

AppAppWorks commented:
Updated, please have a look.

ahoppen reviewed (approved):
Thank you @AppAppWorks. Looks great!

ahoppen commented:
@swift-ci Please test

AppAppWorks commented:
@ahoppen I've found that some of my code comments are misleading and the test case for missing both identifier and type is missing.

I've rewritten the comments and added a new test as follows, if you also think they're necessary and good, I'll force-commit again.

```swift
let names = self.parseParameterNames()
var colon = self.consume(if: .colon)
// try to parse the type regardless of the presence of the preceding colon
// to tackle any unnamed parameter or missing colon
// e.g. [X], (:[X]) or (x [X])
let canParseType = withLookahead { $0.canParseType() }
let type: RawTypeSyntax?
if canParseType {
  type = self.parseType(misplacedSpecifiers: misplacedSpecifiers)
  if colon == nil {
    // mark the preceding colon as missing if the type is present
    // e.g. [X] or (x [X])
    colon = missingToken(.colon)
  }
} else if colon != nil {
  // mark the type as missing if the preceding colon is present
  // e.g. (:) or (_:)
  type = RawTypeSyntax(RawMissingTypeSyntax(arena: self.arena))
} else {
  type = nil
}
```

```swift
assertParse(
  """
  test { (1️⃣:2️⃣) in }
  """,
  diagnostics: [
    DiagnosticSpec(
      locationMarker: "1️⃣",
      message: "expected identifier in parameter",
      fixIts: ["insert identifier"]
    ),
    DiagnosticSpec(
      locationMarker: "2️⃣",
      message: "expected type in parameter",
      fixIts: ["insert type"]
    ),
  ],
  fixedSource: """
    test { (<#identifier#>: <#type#>) in }
    """
)
```

ahoppen commented:
Force push your improved changes and I’ll review them again.

AppAppWorks commented:
Pushed, please check.

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test
